lib/Dialect/FIRRTL/Ops.cpp

```diff
@@ -74,6 +75,116 @@ static LogicalResult verifyCircuitOp(CircuitOp &circuit) {
     return failure();
   }
 
+  // Store a mapping of defname to either the first external module
+  // that defines it or, preferentially, the first external module
+  // that defines it and has no parameters.
+  llvm::DenseMap<Attribute, FExtModuleOp> defnameMap;
+
+  // Verify external modules.
+  for (auto &op : *circuit.getBody()) {
+    auto extModule = dyn_cast<FExtModuleOp>(op);
+    if (!extModule)
+      continue;
+
+    auto defname = extModule.defnameAttr();
+    if (!defname)
+      continue;
+
+    // Check that this extmodule's defname does not conflict with
+    // the symbol name of any module.
+    auto collidingModule = circuit.lookupSymbol(defname.getValue());
+    if (isa_and_nonnull<FModuleOp>(collidingModule)) {
+      auto diag =
+          op.emitOpError()
+          << "attribute 'defname' with value " << defname
+          << " conflicts with the name of another module in the circuit";
+      diag.attachNote(collidingModule->getLoc())
+          << "previous module declared here";
+      return failure();
+    }
+
+    // Find an optional extmodule with a defname collision. Update
+    // the defnameMap if this is the first extmodule with that
+    // defname or if the current extmodule takes no parameters and
+    // the collision does. The latter condition improves later
+    // extmodule verification as checking against a parameterless
+    // module is stricter.
+    FExtModuleOp collidingExtModule;
+    {
+      auto &value = defnameMap[defname];
+      if (value) {
+        collidingExtModule = value;
+        if (value.parameters() && !extModule.parameters())
+          value = extModule;
+      } else {
+        value = extModule;
+      }
+    }
+
+    // Go to next extmodule if no extmodule with the same defname
+    // was found.
+    if (!collidingExtModule)
+      continue;
```

Comment:
would it make sense to pull this into the 'if' above?  Something like: `if (!value) { value = extModule; continue }`.  This eliminates the "else" by making it be the fall through.

Just my personal style, but I wouldn't worry about adding the extra scope around the "value" reference.  Once you do that, you can move the definition of collidingExtModule lower.

Reply:
Yeah, it's cleaner to just move the `continue` up. I can also remove the scope guarding of `value` by just putting that in the `if` statement.